### PR TITLE
feat: add Boulder traffic cams

### PR DESCRIPTION
**Pull request by bbazyk**
boulder_traffic_cams

Ticket to add support for Boulder Traffic Cams, id : #178271931
Boulder cams are livestreams, not images as with the Providence cams. This PR adds react hls player library to enable displaying these streams. Also adds a field to the database which determines how to display the feed.
Cams are visible after 4pm and can be viewed by setting the env field for CITY_NAME to "Boulder" - as was the case in viewing cams before.

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -102,39 +103,77 @@
...
 
 TrafficCam.find_or_initialize_by(
   title: "Tobey Street"
 ) do |r|
   r.url = "http://www.dot.ri.gov/img/travel/camimages/6-2%20Rt%206%20E%20%20%2010%20N%20@%20Tobey%20St.jpg"
   r.location = providence
+  r.feed_format = "image"
+end.save!
+
+TrafficCam.find_or_initialize_by(
```
Comment (jennyckaplan) [inline db/seeds.rb:149]: Let's make adding the Boulder cams a separate commit

File: package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
     "react-apollo": "^2.1.4",
     "react-dom": "^16.8.6",
     "react-highlight-words": "^0.16.0",
+    "react-hls-player": "^3.0.1",
```
Comment (jennyckaplan) [inline package.json:34]: This should also be part of the second commit, since the `ReactHlsPlayer` is only used in the next commit with the front-end changes

File: yarn.lock [hunk 3]
```diff
@@ -4551,7 +4551,7 @@ etag@~1.8.1:
   resolved "https://registry.yarnpkg.com/etag/-/etag-1.8.1.tgz#41ae2eeb65efa62268aebfea83ac7d79299b0887"
```
Comment (jennyckaplan) [inline yarn.lock:4551]: The `yarn.lock` file change should also be in the next commit with the `package.json` change.

**Comment by tmsavage** [comment]
I recognize all of those camera locations haha!

File: db/seeds.rb [hunk 4]
```diff
@@ -7,13 +7,14 @@
 end
 providence.save!
 
-Location.find_or_initialize_by(
+boulder = Location.find_or_initialize_by(
```
Comment (tmsavage) [inline db/seeds.rb:10]: Does this initialize to a boulder variable? In general, should this be done for the other towns?

Comment (bbazyk) [reply db/seeds.rb:10]: The variable is only referenced by initialized "TrafficCam"  objects - just as a means of differentiating between which should be displayed to the widget.

**Review by chewnoill — DISMISSED** [review]
Nice work @bbazyk!

**Comment by jennyckaplan** [comment]
LGTM, except for the `Notice` that still only says `dot.ri.gov` even when displaying Boulder - we should add this in as a field in the database as well and dynamically display the correct source @bbazyk

File: app/javascript/components/widgets/traffic/panel.jsx [hunk 5]
```diff
@@ -84,18 +91,19 @@ const Traffic = () => (
...
-      const [firstColumn, secondColumn] = splitAt(3, trafficCams);
+      const columnSize = Math.floor(trafficCams.length / 2);
+      const [firstColumn, secondColumn] = splitAt(columnSize, trafficCams);
 
       return (
         <>
           <Row>
             <Column>
-              {take(3, firstColumn).map(({ id, ...cam }) => (
+              {take(columnSize, firstColumn).map(({ id, ...cam }) => (
```
Comment (chewnoill) [inline app/javascript/components/widgets/traffic/panel.jsx:101]: does take make sense if we're using the whole list?
can this just be `firstColumn.map(...`?

Comment (chewnoill) [reply app/javascript/components/widgets/traffic/panel.jsx:101]: Oh I see, we want both columns to be the same size

File: db/seeds.rb [hunk 6]
```diff
@@ -109,6 +109,7 @@
   r.url = "http://www.dot.ri.gov/img/travel/camimages/95-21%20I-95%20S%20@%20Broad%20St%20.jpg"
   r.location = providence
   r.feed_format = "image"
+  r.source_url = "http://www.dot.ri.gov"
```
Comment (chewnoill) [inline db/seeds.rb:112]: I don't think we need an extra column for this, if it's just the base url for the url we are already keeping. You can parse this from the js side like this:

```js
const url = "http://www.dot.ri.gov/img/travel/camimages/95-21%20I-95%20S%20@%20Broad%20St%20.jpg"
const source_url = new URL(url).origin
```

File: app/javascript/components/widgets/traffic/panel.jsx [hunk 7]
```diff
@@ -93,6 +94,7 @@ const Traffic = () => (
 
       const columnSize = Math.floor(trafficCams.length / 2);
       const [firstColumn, secondColumn] = splitAt(columnSize, trafficCams);
+      const sourceUrls = uniq(trafficCams.map(cam => cam.sourceUrl));
```
Comment (chewnoill) [inline app/javascript/components/widgets/traffic/panel.jsx:97]:
```js
const sourceUrls = uniq(trafficCams.map(cam => new URL(cam.url).origin));
```

Comment (cbarber) [reply app/javascript/components/widgets/traffic/panel.jsx:97]: Bonus points for functional composition:
```
const getUniqueUrls = compose(uniq, map(cam => new URL(cam.url).origin);
...
          {getUniqUrls(trafficCams).map(url => {
            <Notice>{url}</Notice>
          ))}
```

File: db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class AddFeedFormatToTrafficCams < ActiveRecord::Migration[5.2]
+  def change
+    add_column :traffic_cams, :feed_format, :string
```
[the remark this answers is not in the document]
Comment (chewnoill) [reply db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb:3]: does `default` need to be added before `null: false` or can it be same time? A little confusing about how rails runs migrations, is there a way to see what SQL ends up running here?

**Review by chewnoill — APPROVED** [review]
🚀

File: db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (cbarber) [reply db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb:3]: `null: false` requires a `default: <value>` or existing data. The latter can be accomplished with:
```
add_column :traffic_cams, :feed_format, :string
execute "UPATE traffic_cams SET feed_format = '...'"
change_column :traffic_cams, :feed_format, :string, null: false
```
If you want to use a modal to build the update statement, you create a temporary modal class inside the migration to prevent breaking the migration with future modal changes.

In this particular case, I think `null: false, default: 'image'` is sufficient.

File: app/javascript/components/widgets/traffic/panel.jsx (hunk 7 above)
Comment (bbazyk) [reply app/javascript/components/widgets/traffic/panel.jsx]: it just makes it easier for anyone to export the function later, if they need its output

File: db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (bbazyk) [reply db/migrate/20210527193617_add_feed_format_to_traffic_cams.rb]: This is ready for re-review @cbarber @chewnoill

File: app/javascript/components/widgets/traffic/panel.jsx [hunk 9]
```diff
@@ -90,9 +90,14 @@ const Traffic = () => (
...
       }
-
       const columnSize = Math.floor(trafficCams.length / 2);
       const [firstColumn, secondColumn] = splitAt(columnSize, trafficCams);
+      const getUniqUrls = compose(
+        uniq,
+        map(cam => new URL(cam.url).origin),
+      );
+
+      console.log(getUniqUrls(trafficCams));
```
Comment (chewnoill) [inline app/javascript/components/widgets/traffic/panel.jsx:100]: 🦅

**Review by chewnoill — APPROVED** [review]
there is an extra console log, but otherwise LGTM

Nice commit history also, very clean :atom:
